Cell Charger Display Complete

StenBone opened the pull request:
<!-- The notes within these arrows are for you but can be deleted. -->

## Cell Charger display will now show the current power level of the attached power cell

<!-- Provide a general summary of your change here and in the title. -->
Completes #399
Modified the CellCharger.cs to switch out the emissive maps on the cell charger depending on the power level of the power cell it is charging. 
<!-- Follow with a more concise explanation of your change here. -->
Cell charger display will now emit one of 4 lights depending on power level of the attached power cell.
<!-- What features does this change include/not include? -->

## Pictures/Videos (optional)
<!-- Include photos or videos if possible to help reviewers. -->
[Video demonstrating feature](https://youtu.be/pX0tFG0ipgE)
<!-- It may also be used in our monthly devblog. -->

## Changes to Files

<!-- List any major asset/script/scene changes and why/how they were changed. -->
- SS3D\Assets\Content\Furniture\Machines\Science\CellCharger.cs - modified script to account for new emissive maps
- SS3D\Assets\Content\Furniture\Machines\Science\CellCharger.prefab - modified prefab to account for new emissive material and script changes
- SS3D\Assets\Content\World\CellChargerEmission.mat - new emissive material
- SS3D\Assets\Content\Scenes\Maps\MiniStation.unity - Added cell charger and power cell to scene for testing purposes

Added following emissive map textures for cell charger:
- SS3D\Assets\Art\Textures\Furniture\Machines\Science\Cell Charger\CellChargerOffEmissiveMap.png
- SS3D\Assets\Art\Textures\Furniture\Machines\Science\Cell Charger\CellChargerLowEmissiveMap.png
- SS3D\Assets\Art\Textures\Furniture\Machines\Science\Cell Charger\CellChargerMidLowEmissiveMap.png.meta
- SS3D\Assets\Art\Textures\Furniture\Machines\Science\Cell Charger\CellChargerMidHighEmissiveMap.png
- SS3D\Assets\Art\Textures\Furniture\Machines\Science\Cell Charger\CellChargerHighEmissiveMap.png

## Technical Notes (optional)

<!-- Provide a more technical description of your changes to help save the reviewers some time. -->

## Known issues

<!-- List ANYTHING not working correctly, either part of your new change, or another part of the game. -->
Preexisting issue. The power cell doesn't display the correct orientation to the wires when attached to the cell charger.
<!-- Any known bugs will likely require sorting out before the PR is merged. -->

## Fixes (optional)

<!-- List any issues or other PRs connected to this one. -->
<!-- If this PR CLOSES any issues/PRs, add "Closes" before the number (e.g. "Closes #123"). -->
Closes #399

StenBone commented:
Requesting Review @Centcom

cosmiccoincidence reviewed (changes requested):
Thanks for the contribution and Sorry for the delayed review StenBone.

Some things I noticed and probably should have been made more clear in the original issue (the reason they weren't is that it is an older issue and we had fewer design decisions made at that time).

- Part of our optimization method is to reduce the number of textures as much as possible and the primary way we are achieving this is by using the same palette textures everywhere we can. These palettes can be found in /Assets/Art/Textures/. We also already have an emission material for the palette in /Assets/Content/World/.

This should be used instead of creating a new .png texture for each state of this machine (and subsequently all machines). I think the best way to go about this is to separate the lights on the model itself as individual objects, then using a method similar to the flashlights, toggle the material of the lights.

I can the edit model and set up the prefab again so you don't have to (just let me know).

- Secondly is how the lights are displayed (again, I should have been more clear on this). Have the lights turn on as if it is a meter filling up.
![lights](https://user-images.githubusercontent.com/41941008/123666282-109c9500-d807-11eb-9e88-e847bbbb055e.png)

StenBone commented:
> 
> 
> Thanks for the contribution and Sorry for the delayed review StenBone.
> 
> Some things I noticed and probably should have been made more clear in the original issue (the reason they weren't is that it is an older issue and we had fewer design decisions made at that time).
> 
>     * Part of our optimization method is to reduce the number of textures as much as possible and the primary way we are achieving this is by using the same palette textures everywhere we can. These palettes can be found in /Assets/Art/Textures/. We also already have an emission material for the palette in /Assets/Content/World/.
>       This should be used instead of creating a new .png texture for each state of this machine (and subsequently all machines). I think the best way to go about this is to separate the lights on the model itself as individual objects, then using a method similar to the flashlights, toggle the material of the lights.
>       I can the edit model and set up the prefab again so you don't have to (just let me know).
> 
>     * Secondly is how the lights are displayed (again, I should have been more clear on this). Have the lights turn on as if it is a meter filling up.
>       ![lights](https://user-images.githubusercontent.com/41941008/123666282-109c9500-d807-11eb-9e88-e847bbbb055e.png)

Thank you so much for the reply @cosmiccoincidence! Much appreciated. 
Using shared textures does make sense and explains why my approach wasn't used in the beginning. 
I'll use the new light sequence that you displayed in your picture.
Please update the model and prefab. I don't have much skill in 3D modelling applications, so that would be a big help. 
I'll study how the flashlight works and revise my solution once you have completed the model updates.

cosmiccoincidence commented:
Updated the cell charger object so the lights are separate objects themselves, and updated the prefab as well.

Good luck! and sorry about the lack of a specific design in the original issue.

cosmiccoincidence commented:
I'm going to add `[WIP]` to the title. Please rename the title again, removing the WIP when you are ready for another review.

Alainx277 commented on Assets/Content/Furniture/Machines/Science/CellCharger.cs:
```diff
@@ -49,10 +54,58 @@ private void Recharge()
+    /// </summary>
+    private void TurnOffLights() 
+    {
+        redLight.TurnLightOff();
+        orangeLight.TurnLightOff();
+        yellowLight.TurnLightOff();
+        greenLight.TurnLightOff();
```
I think it would make a lot of the code shorter if you create an array of all the lights (like `lights = new[] { redLight, orangeLight, yellowLight, greenLight }`.
Then you can just call `TurnLightOff` for each element.

Ryan089 replied:
Recommendation supported, but left for consideration for future work.

Alainx277 commented on Assets/Content/Furniture/Machines/Science/CellCharger.cs:
```diff
@@ -49,10 +54,58 @@ private void Recharge()
+    /// <param name="powerPercentage">Value between 0 - 1</param>
+    private void UpdateLights(float powerPercentage) 
+    {
+        if (0 <= powerPercentage && powerPercentage < .25f) // power @ 25%
+        {
+            redLight.MakeLightBlink();
+        }
+        else if (.25f <= powerPercentage && powerPercentage < .50f) // power @ 50%
+        {
+            redLight.MakeLightStayOn();
+            orangeLight.MakeLightBlink();
+        }
+        else if (.50f <= powerPercentage && powerPercentage < .75f) // power @ 75%
+        {
+            redLight.MakeLightStayOn();
+            orangeLight.MakeLightStayOn();
+            yellowLight.MakeLightBlink();
+        }
+        else if  (.75f <= powerPercentage && powerPercentage < 1f)
+        {
+            redLight.MakeLightStayOn();
+            orangeLight.MakeLightStayOn();
+            yellowLight.MakeLightStayOn();
+            greenLight.MakeLightBlink();
+        }
+        else if (1f <= powerPercentage)
+        {
+            redLight.MakeLightStayOn();
+            orangeLight.MakeLightStayOn();
+            yellowLight.MakeLightStayOn();
+            greenLight.MakeLightStayOn();
         }
```
By using the lights array, this could be simplified by taking the increment (`1f / lights.Length`) and then calculating the index of the blinking light. This tells you which lights should be on (those before the blinking light) and off too (those after the blinking light).

StenBone replied:
I got this working locally, but couldn't figure out to incorporate it with the networking changes from @Ryan089. The changes are pushed up to the PR for anyone wanting to rewrite  the cell charger to use an array of lights.

Ryan089 replied:
Recommendation supported, but left for consideration for future work.

Alainx277 commented on Assets/Content/Items/Functional/Parts/Lights/BlinkingLight.cs:
```diff
@@ -0,0 +1,75 @@
+    {
+        lightMesh = GetComponent<MeshRenderer>();
+        TurnLightOff();
+        blinkCoroutine = StartBlinking();
```
I would use the `Coroutine` object returned by `StartCoroutine` to turn it off again, instead of creating an Enumerator that does nothing at the start.

StenBone replied:
TBH, I'm not the most familiar with coroutines.

Ryan089 replied:
Recommendation supported, but left for consideration for future work.

Alainx277 commented on Assets/Content/Items/Functional/Parts/Lights/BlinkingLight.cs:
```diff
@@ -0,0 +1,75 @@
+    [Tooltip("How long should the light stay on when blinking. Must be less than blink period.")]
+    [SerializeField] private float blinkLength = .5f;
+    
+    void Awake() 
```
Is there a reason why this is in `Awake` and not `Start`?

StenBone replied:
I was testing it locally in another project and left that there. No reason. Switching it back to Start.

Alainx277 commented on Assets/Content/Items/Functional/Parts/Lights/BlinkingLight.cs:
```diff
@@ -0,0 +1,75 @@
+    [SerializeField] private Material emissiveMaterial;
+
+    [Header("Blink Settings")]
+    [Tooltip("How often should a blink occur in seconds.")]
```
Isn't this how many seconds between blinks?

StenBone replied:
Very true. I updated the tooltip with clearer language.

StenBone commented:
Removed WIP. 
Fetched and merged most recent changes into PR. Implemented @Ryan089's network changes and followed suggestions from @Alainx277. I was unable to convert the functionality over to using an array. Can take another stab at it if that is required to pass review.

Ryan089 reviewed (changes requested):
Thanks for your work so far. There is just one (very simple) thing that I'd like to see before this is merged. Please open MainStation, save it, commit the changes and push it up.

When working with NetworkIdentities, some changes to prefabs require each scene that they are in to be resaved, so that Mirror can regenerate the Scene ID. If that doesn't happen, attempting to build the executable will result in something like this:

![image](https://user-images.githubusercontent.com/64360820/127110212-3c40bd86-5e1a-4234-97e3-9572dca2c7af.png)

Once this is done, this should be good to merge.

StenBone commented:
> Thanks for your work so far. There is just one (very simple) thing that I'd like to see before this is merged. Please open MainStation, save it, commit the changes and push it up.
> 
> When working with NetworkIdentities, some changes to prefabs require each scene that they are in to be resaved, so that Mirror can regenerate the Scene ID. If that doesn't happen, attempting to build the executable will result in something like this:
> 
> ![image](https://user-images.githubusercontent.com/64360820/127110212-3c40bd86-5e1a-4234-97e3-9572dca2c7af.png)
> 
> Once this is done, this should be good to merge.

Will do, appreciate the feedback @Ryan089!

StenBone commented:
@Ryan089 Just reopened Main Station and saved the changes. Pushed changes up to master.

StenBone commented:
Also added @Ryan089 and @Alainx277  as collaborators.
![image](https://user-images.githubusercontent.com/4539982/127238626-4f22eed3-73ff-4ceb-889a-f2d537a35ead.png)

Ryan089 reviewed (approved):
Congratulations on completing your first issue for this project. Alainx's suggestions would make the code more elegant and maintainable, and it may be something that you or another contributor could return to in the future if desired. However, the functional requirements have been met and I'm happy to merge. Great work!
